### ISSUE TITLE
XGBoostLSS for uncertainty estimation in binary classification

### ISSUE THREAD

**Issue by mathijs02** [StatMixedML/XGBoostLSS#10]
Hi,

Thanks for creating the XGBoostLSS (and LightGBMLSS) package, this an important extension to ML algorithms that only provide point estimates for the mean.

I have a question, or thought, about capturing uncertainty for binary classification. I would be interested in estimating the uncertainty in the model score in a binary classification, as a consequence of the amount of data that the algorithm has seen during training. If in a certain part of the feature space there was one positive and one negative training instance, the point estimate will be *p*=0.5, but with a lot less certainty than if there would have been a hundred positive and a hundred negative instances. One way to estimate the uncertainty in the binomial proportion *p* for a given number of negative (*a*) and positive (*b*) observations is to use a Beta distribution with parameters *α* = *a* + 1/2 and *β* = *b* + 1/2. This is referred to as Jeffreys interval in this paper, for example:
https://projecteuclid.org/journals/statistical-science/volume-16/issue-2/Interval-Estimation-for-a-Binomial-Proportion/10.1214/ss/1009213286.full

I wonder if the methods of XGBoostLSS can be used, or extended, to return the *α, β* parameters of a Beta distribution that fits the binary data observed during training. This would be a very valuable estimate of uncertainty. I see that the Beta distribution is implemented in XGBoostLSS, but if I understand correctly, this is to fit a regression on target variables that themselves follow the Beta distribution (not for classification of binomial target variables). And I'm not sure how to obtain the NLL for the Beta distribution for a binary target variable: the Beta PDF is 0 or ∞ at *x*=0 and *x*=1.

I have to admit that I am a bit out of my depth on the mathematics of this question, so if my reasoning is entirely in the wrong direction or incompatible with the idea behind XGBoostLSS, I am also happy to hear!

**Comment by StatMixedML** [comment]
@mathijs02 Thanks for posting this very interesting question!

> I see that the Beta distribution is implemented in XGBoostLSS, but if I understand correctly, this is to fit a regression on target variables that themselves follow the Beta distribution (not for classification of binomial target variables). 

This is correct. So far, the beta is implemented only if the response variable ```y``` is beta distributed. In principle, for any parametric distribution with parameter vector ß, XGBoostLSS is able to estimate and return all parameters. Hence, we need to make sure we can transform your problem into a suitable distribution with log-likelihood density that is used for estimating the distributional parameters. Not sure if I am correct, but I might have seen somewhere transforming a Bernoulli variable into a beta-distributed variable. But let me go through the paper you've shared and come back to you.

**Comment by mathijs02** [comment]
> Hence, we need to make sure we can transform your problem into a suitable distribution with log-likelihood density that is used for estimating the distributional parameters, even though XGBoostLSS is developed mainly for regression tasks.

That was indeed my initial thought. The more I think of it, the less this line of reasoning seems to go anywhere. The issue, I think, is that the likelihood for classification with a Beta distribution is calculated on a set of instances. This is in contrast with other common loss functions, which are calculated on a per-instance basis. For example, if in a leaf there is a set of negative (*a*) and positive (*b*) observations, we can calculate the likelihood for *x* = *b*/(*a* + *b*), given a Beta function with distributional parameters *α*, *β*. However, the values for *x*=0 and *x*=1 for a Beta function do not give useful likelihoods in this case, so the likelihood makes no sense on single instances.

> Maybe Bayesian models are suited for your task, since the posterior distribution allows to calculate intervals etc. directly. 

I think you are right. My reasoning was to make a Bayesian model out of XGBoost/LightGBM, because they give such high performance on tabular data. I will try to find if there is any more literature on this.
